core/pom.xml

```diff
@@ -66,7 +66,13 @@
     </dependency>
     <dependency>
       <groupId>org.apache.hadoop</groupId>
-      <artifactId>hadoop-client</artifactId>
+      <artifactId>${hadoop-client-api.artifact}</artifactId>
```

Comment:
@sunchao @dongjoon-hyun I find there are some bad case with using variables here:

- if we run `mvn dependency:tree -pl resource-managers/yarn -Phadoop-2.7 -Pyarn -am`, the dependency-tree is correct
- if we run `mvn dependency:tree -pl resource-managers/yarn -Phadoop-2.7 -Pyarn`, the dependency-tree is wrong, the dependency of hadoop 3.3.1 exists in the dependency tree, even though we have added the profile of - Phadoop2.7

![image](https://user-images.githubusercontent.com/1475305/133629341-78ac4158-39f5-4bd3-af57-607c10ea49db.png)


So if we execute 
```
mvn clean install -DskipTests -pl resource-managers/yarn -am -Phadoop-2.7 -Pyarn
mvn test -pl resource-managers/yarn -Phadoop-2.7 -Pyarn -DwildcardSuites=org.apache.spark.deploy.yarn.YarnClusterSuite
```

The test case `YarnClusterSuite` will failed as follows:
```
Discovery starting.
Discovery completed in 277 milliseconds.
Run starting. Expected test count is: 1
YarnClusterSuite:
*** RUN ABORTED ***
  java.lang.NoClassDefFoundError: org/apache/hadoop/shaded/com/google/inject/servlet/ServletModule
  at java.lang.ClassLoader.defineClass1(Native Method)
  at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
  at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
  at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
  at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
  at java.security.AccessController.doPrivileged(Native Method)
  at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
  ...
  Cause: java.lang.ClassNotFoundException: org.apache.hadoop.shaded.com.google.inject.servlet.ServletModule
  at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
  at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
  at java.lang.ClassLoader.defineClass1(Native Method)
  at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
  at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
  at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
  at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
  at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
  ...

```

Reply:
We have to add the `-am` to test the `yarn` module separately to ensure that the dependency is correct
```
mvn test -pl resource-managers/yarn -Phadoop-2.7 -am -Pyarn -DwildcardSuites=org.apache.spark.deploy.yarn.YarnClusterSuite -Dtest=none
```
```
Run completed in 5 minutes, 7 seconds.
Total number of tests run: 27
Suites: completed 2, aborted 0
Tests: succeeded 27, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

Reply:
Interesting. Let me take a look. Thanks for reporting it @LuciferYang !

Reply:
This behavior looks expected to me... If you specify `-pl resource-managers/yarn` without the `-am` flag, then you're explicitly telling Maven _not_ to rebuild the `core` module. If the `core` module has previously been built with Hadoop 3.3.1, then you will just use it as-is with the old dependencies. IMO, this is just an example of how `-pl` without `-am` needs to be used with care.

Reply:
Hi, @LuciferYang . I agree with @xkrogen .

When you use `--am`, the other module is the published SNAPSHOT poms and jars which is built with the default (Hadoop 3) configuration.

Reply:
> When you use `--am`, the other modules are the published SNAPSHOT poms and jars which is built with the default (Hadoop 3) configuration.

@dongjoon-hyun I think you mean when you _don't_ use `--am`? (just want to avoid any confusion here)

Reply:
Oh, my. Yes, it was a typo. Yes, 'you use without `--am`' instead of 'you use `--am`'.

Reply:
I'll revise it.

Reply:
What I'm executing now are
```
mvn clean install -DskipTests -pl resource-managers/yarn -am -Phadoop-2.7 -Pyarn
mvn test -pl resource-managers/yarn -Phadoop-2.7 -Pyarn -DwildcardSuites=org.apache.spark.deploy.yarn.YarnClusterSuite
```

So we can't `mvn test` the `yarn` module independently with `hadoop-2.7` now? Or should use other execution commands?

Reply:
@LuciferYang . To achieve what you want, it seems that we need to do cross-publishing like Scala 2.12 and 2.13. But, it will not happen for Hadoop because Hadoop have too many versions.

Reply:
Oh, you are trying to `install` first?

Reply:
Yes，I do install first～ I think it was passed before，for example, in Spark 3.1, I will double check it tomorrow.

It's really troublesome to reply on mobile phone，haha

Reply:
Ahh ... Yes, I see. It does seem that since you just did `install -Phadoop-2.7 -pl resource-managers/yarn -am`, this should work properly. The installed version of `spark-core` should be the one with Hadoop 2.7 dependencies.

Reply:
```
mvn clean install -DskipTests -pl resource-managers/yarn -am -Phadoop-2.7 -Pyarn
mvn test -pl resource-managers/yarn -Phadoop-2.7 -Pyarn -DwildcardSuites=org.apache.spark.deploy.yarn.YarnClusterSuite
```
I check the above commands in master, branch-3.2 and branch-3.1:

- master: failed
- branch-3.2 : failed
- branch-3.1: success

Reply:
Sorry forgot about this. @LuciferYang just to recap, do you still see a bug here? or the behavior is expected.

Reply:
@sunchao Yes, this problem still exists, only behavior of  branch-3.1 is expected at present

Reply:
I test these command in 3.2-rc4(3.2-rc5 can't build with hadoop-2.7 now) , the problem still exists

Reply:
@LuciferYang could you check with the fix in #34100? I just tested it with the command you pasted above:
```
mvn clean install -DskipTests -pl resource-managers/yarn -am -Phadoop-2.7 -Pyarn
mvn test -pl resource-managers/yarn -Phadoop-2.7 -Pyarn -DwildcardSuites=org.apache.spark.deploy.yarn.YarnClusterSuite
```
and the tests all passed for me.

Reply:
@sunchao Yes, the behavior is  expected now ~ thx ~